### PR TITLE
fix bpf prog double load 

### LINKED ISSUE #700: After kmesh restart， bpf progs with same name double
By delete kmesh pod, and then we can see bpf prog doubled.
```
bpftool prog list |grep sock
12946: cgroup_sock_addr  name cgroup_connect4  tag c0b238fa37ac74c6  gpl
12947: cgroup_sock_addr  name cgroup_connect6  tag 894a81446c6a97fe  gpl
12948: sock_ops  name sockops_prog  tag 75dc4653e3233520  gpl
13016: cgroup_sock_addr  name cgroup_connect4  tag c0b238fa37ac74c6  gpl
13017: cgroup_sock_addr  name cgroup_connect6  tag 894a81446c6a97fe  gpl
13018: sock_ops  name sockops_prog  tag 75dc4653e3233520  gpl
```

### PR DESCRIPTION
**Pull request by lec-bit**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
After restarting, multiple bpf_prog will be loaded. Fix this point and replace it with dynamic update.
**Which issue(s) this PR fixes**:
Fixes #700

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/bpf/bpf_kmesh_workload.go [hunk 1]
```diff
@@ -135,34 +127,55 @@ func (sc *BpfSockConnWorkload) close() error {
...
 	if GetStartType() == Restart {
-		return err
-	}
+		pinPath4 := filepath.Join(sc.Info.BpfFsPath, "sockconn_prog")
+		sclink4, err := link.LoadPinnedLink(pinPath4, &ebpf.LoadPinOptions{})
+		if err != nil {
+			return err
+		}
+		if err = sclink4.Update(cgopt4.Program); err != nil {
+			return fmt.Errorf("updating link %s %w", pinPath4, err)
```
Comment (LiZhenCheng9527) [inline pkg/bpf/bpf_kmesh_workload.go:149]:
```suggestion
			return fmt.Errorf("updating link %s failed: %w", pinPath4, err)
```

File: pkg/bpf/bpf_kmesh_workload.go [hunk 2]
```diff
@@ -135,34 +127,55 @@ func (sc *BpfSockConnWorkload) close() error {
...
-	}
-	if err := sc.Link6.Pin(sc.Info.BpfFsPath + "sockconn6_prog"); err != nil {
-		return err
+		pinPath6 := filepath.Join(sc.Info.BpfFsPath, "sockconn6_prog")
+		sclink6, err := link.LoadPinnedLink(pinPath6, &ebpf.LoadPinOptions{})
+		if err != nil {
+			return err
+		}
+		if err = sclink6.Update(cgopt6.Program); err != nil {
+			return fmt.Errorf("updating link %s %w", pinPath6, err)
```
Comment (LiZhenCheng9527) [inline pkg/bpf/bpf_kmesh_workload.go:158]: ditto

File: pkg/bpf/bpf_kmesh_workload.go [hunk 3]
```diff
@@ -135,34 +127,54 @@ func (sc *BpfSockConnWorkload) close() error {
...
-
 	if GetStartType() == Restart {
-		return err
-	}
+		pinPath4 := filepath.Join(sc.Info.BpfFsPath, "sockconn_prog")
+		sclink4, err := link.LoadPinnedLink(pinPath4, &ebpf.LoadPinOptions{})
+		if err != nil {
+			return err
+		}
+		if err = sclink4.Update(cgopt4.Program); err != nil {
```
Comment (YaoZengzeng) [inline pkg/bpf/bpf_kmesh_workload.go:148]:
```suggestion
		if err := sclink4.Update(cgopt4.Program); err != nil {
```
just a convention

File: pkg/bpf/bpf_kmesh_workload.go [hunk 4]
```diff
@@ -285,19 +288,27 @@ func (so *BpfSockOpsWorkload) Attach() error {
...
 	if GetStartType() == Restart {
-		return nil
-	}
+		pinPath := filepath.Join(so.Info.BpfFsPath, "cgroup_sockops_prog")
+		solink, err := link.LoadPinnedLink(pinPath, &ebpf.LoadPinOptions{})
+		if err != nil {
+			return err
+		}
+		if err = solink.Update(cgopt.Program); err != nil {
+			return fmt.Errorf("updating link %s %w", pinPath, err)
```
Comment (YaoZengzeng) [inline pkg/bpf/bpf_kmesh_workload.go:298]:
```suggestion
			return fmt.Errorf("updating link %s failed: %w", pinPath, err)
```

File: pkg/bpf/bpf_kmesh_workload.go [hunk 5]
```diff
@@ -135,34 +127,54 @@ func (sc *BpfSockConnWorkload) close() error {
...
+		if err := sclink4.Update(cgopt4.Program); err != nil {
+			return fmt.Errorf("updating link %s failed: %w", pinPath4, err)
+		}
 
-	if err := sc.Link.Pin(sc.Info.BpfFsPath + "sockconn_prog"); err != nil {
-		return err
-	}
-	if err := sc.Link6.Pin(sc.Info.BpfFsPath + "sockconn6_prog"); err != nil {
-		return err
+		pinPath6 := filepath.Join(sc.Info.BpfFsPath, "sockconn6_prog")
```
Comment (hzxuzhonghu) [inline pkg/bpf/bpf_kmesh_workload.go:152]: can you define a const, and prevent hardcoding

File: pkg/bpf/bpf_kmesh_workload.go [hunk 6]
```diff
@@ -135,34 +127,54 @@ func (sc *BpfSockConnWorkload) close() error {
...
-
 	if GetStartType() == Restart {
-		return err
-	}
+		pinPath4 := filepath.Join(sc.Info.BpfFsPath, "sockconn_prog")
+		sclink4, err := link.LoadPinnedLink(pinPath4, &ebpf.LoadPinOptions{})
+		if err != nil {
+			return err
+		}
+		if err := sclink4.Update(cgopt4.Program); err != nil {
```
Comment (hzxuzhonghu) [inline pkg/bpf/bpf_kmesh_workload.go:148]: can we abstract these logic into a helper function, like `bpfProgUpdate` , just a suggestion, may name more appropriately

File: pkg/bpf/bpf_kmesh_workload.go [hunk 7]
```diff
@@ -135,34 +127,54 @@ func (sc *BpfSockConnWorkload) close() error {
...
+		if err := sc.Link.Pin(sc.Info.BpfFsPath + "sockconn_prog"); err != nil {
+			return err
+		}
+
+		sc.Link6, err = link.AttachCgroup(cgopt6)
+		if err != nil {
+			return err
+		}
+
+		if err := sc.Link6.Pin(sc.Info.BpfFsPath + "sockconn6_prog"); err != nil {
```
Comment (hzxuzhonghu) [inline pkg/bpf/bpf_kmesh_workload.go:175]: why donot you use filepath.Join here, please check elsewhere

File: pkg/bpf/bpf_kmesh_workload.go [hunk 8]
```diff
@@ -385,8 +396,25 @@ func (sm *BpfSendMsgWorkload) loadKmeshSendmsgObjects() (*ebpf.CollectionSpec, e
 		return nil, err
 	}
 
+	// sendmsg ebpf prog is mounted on sockmap and processed for each socket.
+	// It has the following characteristics:
+	// 1. Multiple sk_msg ebpf prog can exist at the same time
+	// 2. If the old sk_msg ebpf program is not pinned, it will wait until all
```
Comment (hzxuzhonghu) [inline pkg/bpf/bpf_kmesh_workload.go:402]: i am not sure i understand this feature

File: pkg/bpf/bpf_kmesh_workload.go [hunk 9]
```diff
@@ -385,8 +396,25 @@ func (sm *BpfSendMsgWorkload) loadKmeshSendmsgObjects() (*ebpf.CollectionSpec, e
...
+	// sockets on the old sk_msg ebpf prog are disconnected before automatically detaching.
+	// Therefore, the following methods are used to achieve seamless replacement
+	// 1) loading new sk_msg prog
+	// 2) unpin old psk_msg rog
+	// 3) pin new sk_msg prog
+	// 4) attach new sk_msg prog
```
Comment (hzxuzhonghu) [inline pkg/bpf/bpf_kmesh_workload.go:408]: where is this attach?

Comment (bitcoffeeiux) [inline pkg/bpf/bpf_kmesh_workload.go:408]: Describe the service logic changes after the preceding operations are performed.

File: pkg/bpf/bpf_kmesh_workload.go [hunk 10]
```diff
@@ -385,8 +396,25 @@ func (sm *BpfSendMsgWorkload) loadKmeshSendmsgObjects() (*ebpf.CollectionSpec, e
...
 	if GetStartType() == Restart {
-		return spec, nil
+		oldSkMsg, err := ebpf.LoadPinnedProgram(sm.Info.BpfFsPath+"/sendmsg_prog", nil)
+		if err != nil {
+			log.Errorf("LoadPinnedProgram failed:%v", err)
+		}
+
+		if err = oldSkMsg.Unpin(); err != nil {
```
Comment (bitcoffeeiux) [inline pkg/bpf/bpf_kmesh_workload.go:415]: Does unpinning here unpin the map?

File: pkg/bpf/bpf_kmesh_workload.go [hunk 11]
```diff
@@ -83,15 +84,6 @@ func (sc *BpfSockConnWorkload) loadKmeshSockConnObjects() (*ebpf.CollectionSpec,
 		return nil, err
 	}
 
-	if GetStartType() == Restart {
-		return spec, nil
-	}
-
-	value := reflect.ValueOf(sc.KmeshCgroupSockWorkloadObjects.KmeshCgroupSockWorkloadPrograms)
```
[the remark this answers is not in the document]
Comment (lec-bit) [reply pkg/bpf/bpf_kmesh_workload.go:90]: It is no longer necessary here. Current persistence requires pinning bpf_link, which manages bpf_prog.
If we continue to pin bpf_prog here, you need to process the old prog during the next load.

**Comment by hzxuzhonghu** [comment]
/lgtm

Just one comment left from me https://github.com/kmesh-net/kmesh/pull/763#discussion_r1730503293

**Review by hzxuzhonghu — APPROVED** [review]
/lgtm

File: pkg/bpf/bpf_kmesh_workload.go [hunk 12]
```diff
@@ -385,8 +396,26 @@ func (sm *BpfSendMsgWorkload) loadKmeshSendmsgObjects() (*ebpf.CollectionSpec, e
...
+	// 1) loading new sk_msg prog
+	// 2) unpin old sk_msg prog: If sockmap is deleted, sk_msg will also be cleaned up
+	// 3) pin new sk_msg prog
+	// 4) attach new sk_msg prog(in SendMsg.Attach): Replace the old sk_msg prog
 	if GetStartType() == Restart {
-		return spec, nil
+		pinPath := filepath.Join(sm.Info.BpfFsPath, "sendmsg_prog")
+		oldSkMsg, err := ebpf.LoadPinnedProgram(pinPath, nil)
+		if err != nil {
+			log.Errorf("LoadPinnedProgram failed:%v", err)
```
Comment (hzxuzhonghu) [inline pkg/bpf/bpf_kmesh_workload.go:413]: nit: should make the log more readable, adding `pinPath`

**Pull request merged**